File: engine/julia/weekly_squeeze_engine.jl
# engine/julia/weekly_squeeze_engine.jl
# OptionAlpha Agent — Module Q5: Julia Weekly Squeeze & Heikin Ashi Engine

module WeeklySqueezeEngineJulia

export calculate_heikin_ashi, is_in_squeeze, midpoint_entry

struct HeikinAshiBar
    ha_open::Float64
    ha_high::Float64
    ha_low::Float64
    ha_close::Float64
    is_strong_bull::Bool
    is_strong_bear::Bool
end

function calculate_heikin_ashi(o::Float64, h::Float64, l::Float64, c::Float64, prev_o::Float64, prev_c::Float64)
    ha_open = (prev_o + prev_c) / 2.0
    ha_close = (o + h + l + c) / 4.0
    ha_high = max(h, ha_open, ha_close)
    ha_low = min(l, ha_open, ha_close)
    is_strong_bull = (ha_close > ha_open) && (abs(ha_low - ha_open) < 1e-4)
    is_strong_bear = (ha_close < ha_open) && (abs(ha_high - ha_open) < 1e-4)
    return HeikinAshiBar(ha_open, ha_high, ha_low, ha_close, is_strong_bull, is_strong_bear)
end

function is_in_squeeze(bb_u::Float64, bb_l::Float64, kc_u::Float64, kc_l::Float64)
    return (bb_u < kc_u) && (bb_l > kc_l)
end

function midpoint_entry(o::Float64, c::Float64)
    return (o + c) / 2.0
end

end # module

File: engine/rust/src/weekly_squeeze_engine.rs
// engine/rust/src/weekly_squeeze_engine.rs
// OptionAlpha Agent — Module Q4: Rust Weekly Squeeze & Heikin Ashi Engine

#[derive(Debug, Clone, Copy)]
pub struct HeikinAshiBar {
    pub open: f64,
    pub high: f64,
    pub low: f64,
    pub close: f64,
    pub is_strong_bull: bool,
    pub is_strong_bear: bool,
}

impl HeikinAshiBar {
    pub fn new(o: f64, h: f64, l: f64, c: f64, prev_o: f64, prev_c: f64) -> Self {
        let ha_open = (prev_o + prev_c) / 2.0;
        let ha_close = (o + h + l + c) / 4.0;
        let ha_high = h.max(ha_open.max(ha_close));
        let ha_low = l.min(ha_open.min(ha_close));
        let is_strong_bull = (ha_close > ha_open) && ((ha_low - ha_open).abs() < 1e-4);
        let is_strong_bear = (ha_close < ha_open) && ((ha_high - ha_open).abs() < 1e-4);

        Self {
            open: ha_open,
            high: ha_high,
            low: ha_low,
            close: ha_close,
            is_strong_bull,
            is_strong_bear,
        }
    }
}

pub struct WeeklySqueezeEngineRust;

impl WeeklySqueezeEngineRust {
    pub fn is_in_squeeze(bb_u: f64, bb_l: f64, kc_u: f64, kc_l: f64) -> bool {
        (bb_u < kc_u) && (bb_l > kc_l)
    }

    pub fn midpoint_entry(o: f64, c: f64) -> f64 {
        (o + c) / 2.0
    }
}
